sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuArrowEvalPythonExec.scala

```diff
@@ -430,18 +430,46 @@ class GpuArrowPythonRunner(
 
       protected override def writeIteratorToStream(dataOut: DataOutputStream): Unit = {
         val writer = {
+
+          lazy val flattenColumnNames: (StructField, Boolean) => Seq[(String, Boolean)] =
+            (f, appendMe)=> {
+              val colNames = mutable.ArrayBuffer[(String, Boolean)]()
+              if (appendMe) {
+                colNames.append((f.name, f.nullable))
+              }
+              f.dataType match {
+                case structType: StructType =>
+                  colNames.appendAll(structType.flatMap(field => flattenColumnNames(field, true)))
+                case arrayType: ArrayType =>
+                  // Dig into the child but skip itself by specifying `appendMe` to false
+                  colNames.appendAll(flattenColumnNames(
+                    StructField("", arrayType.elementType, f.nullable), false))
```

Comment:
Why create an entry at all if we're not going to append it?  Seems like we can skip the `appendMe` stuff and just have a recursive method that only reports what names it has.  I think this can be significantly simplified with the observation that only structs trigger names.  For example:
```
private def flattenNames(d: DataType): Seq[String] = d match {
  case s: StructType => s.flatMap(sf => Seq(sf.name) ++ flattenNames(sf.dataType))
  case a: ArrayType => names(a.elementType)
  case m: MapType => names(m.keyType) ++ names(m.valueType)
  case _ => Nil
}
[...]
val columnNames = flattenNames(pythonInSchema)
```

Reply:
Good suggestion, i did a little change to support `nullable` flag, which is being used here when specifying the column names.